code/modules/economy/holopay.dm

```diff
@@ -0,0 +1,249 @@
+	"cross", "cut", "donate", "dna", "flask", "glass-cheers", "glass-martini-alt", "hand-holding-usd", "heart", "heart-broken",
+	"hamburger", "hat-cowboy-side", "money-check-alt", "music", "pizza-slice", "prescription-bottle-alt", "radiation", "robot", "smile",
+	"tram", "trash")
+	/// The brand icon chosen by the user
+	var/shop_logo = "donate"
+
+/obj/machinery/holopay/New(turf/location, mob/living/user, obj/item/card/id/card, datum/bank_account/account)
+	loc = location
+	owner = user
+	linked_card = card
+	linked_account = account
+	desc = "Pays directly into [owner]'s bank account."
+	add_atom_colour("#77abff", FIXED_COLOUR_PRIORITY)
+	set_light(2)
+	visible_message(span_notice("A holographic pay stand appears."))
+	/// Start checking if the owner has left or died
+	RegisterSignal(owner, list(COMSIG_MOVABLE_MOVED, COMSIG_LIVING_STATUS_UNCONSCIOUS), .proc/check_operation)
+	return ..()
+
+/**
+ * A periodic check to see if the owner has left the area, died, or is afk.
+ * Deletes the holopay if any condition is true.
+ */
+/obj/machinery/holopay/proc/check_operation()
+	SIGNAL_HANDLER
+	if(!owner || !owner.client || !isliving(owner) || !IN_GIVEN_RANGE(src, owner, MAX_HOLO_RANGE))
+		QDEL_NULL(linked_card?.my_store)
+		qdel(src)
+
+/obj/machinery/holopay/Destroy()
+	playsound(loc, "sound/effects/empulse.ogg", 40, TRUE)
+	visible_message(span_notice("The pay stand vanishes."))
+	return ..()
+
+/obj/machinery/holopay/examine(mob/user)
+	. = ..()
+	if(force_fee)
+		. += span_boldnotice("This holopay forces a payment of <b>[force_fee]</b> credit\s per swipe instead of a variable amount.")
+
+/obj/machinery/holopay/ui_interact(mob/user, datum/tgui/ui)
+	. = ..()
+	if(!linked_account)
+		balloon_alert(user, "no registered owner")
+		return FALSE
+	ui = SStgui.try_update_ui(user, src, ui)
+	if(!ui)
+		ui = new(user, src, "HoloPay")
+		ui.open()
+
+/obj/machinery/holopay/ui_status(mob/user)
+	. = ..()
+	if(!in_range(user, src) && !isobserver(user))
+		return UI_CLOSE
+
+/obj/machinery/holopay/ui_static_data(mob/user)
+	. = list()
+	.["available_logos"] = available_logos
+	.["description"] = desc
+	.["force_fee"] = force_fee
+	.["max_fee"] = MAX_FEE
+	.["name"] = name
+	.["owner"] = linked_account?.account_holder || null
+	.["shop_logo"] = shop_logo
+
+/obj/machinery/holopay/ui_data(mob/user)
+	. = list()
+	var/obj/item/card/id/id_card
+	if(isliving(user))
+		var/mob/living/card_holder = user
+		id_card = card_holder.get_idcard(TRUE)
+	if(id_card?.registered_account)
+		.["user"] = list()
+		.["user"]["name"] = id_card.registered_account.account_holder
+		.["user"]["balance"] = id_card.registered_account.account_balance
+
+/obj/machinery/holopay/ui_act(action, list/params, datum/tgui/ui)
+	. = ..()
+	if(.)
+		return FALSE
+	switch(action)
+		if("done")
+			ui.send_full_update()
+			return TRUE
+		if("fee")
+			/// Input checks for fee amount
```

Comment:
What does this comment refer to?

Reply:
the ui has a force fee setting and we want to make sure it's a number in valid range here